### PR TITLE
Refactor camera & mic features into MediaManager

### PR DESCRIPTION
**Pull request by pgivel**
## Description

This splits the functionnal and UI usage. It also decorrelates the microphone & camera features, allowing to stream one or the other independantly.

Fixes #<issue>

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [ ] I've read & comply with the [contributing guidelines](https://github.com/Genymobile/genymotion-device-web-player/blob/main/CONTRIBUTING.md)
- [ ] I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [ ] I have made corresponding changes to the documentation (README.md).
- [ ] I've checked my modifications for any breaking changes.

### CODE REVIEW

File: src/plugins/MediaManager.js [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        this.instance.mediaEventsEnabled = true;
+
+        this.videoStreaming = false;
+        this.audioStreaming = false;
+
+        this.videoWidth = videoWidth;
+        this.videoHeight = videoHeight;
+        this.videoWithMicrophone = videoWithMicrophone;
+
+        this.setupPermissions();
```
Comment (skornous) [inline src/plugins/MediaManager.js:32]: is it really necessary to redo a setupPermissions ? if [l.18](https://github.com/Genymobile/genymotion-device-web-player/pull/40/files#diff-97a1a105b0f7db226929204ef391737aac32f2a5cbaf8a20278a55453fb07cddR18) passes it should already have been done so redoing it might double register the `on<Microphone/Camera>PermissionChange` listeners

Comment (pgivel) [reply src/plugins/MediaManager.js:32]: You're right, I forgot to remove this line when adding l.18 :) Thanks!

**Review by jparez — APPROVED** [review]

File: src/plugins/MediaManager.js [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+
+    /**
+     * Find RTCRtpSender corresponding RTCRtpTransceiver and set its direction.
+     *
+     * @param {RTCRtpSender} sender    Peer (sender).
+     * @param {string}       direction Transceiver direction.
+     */
+    setTransceiverDirection(sender, direction) {
+        // find transceiver that contains sender
+        this.instance.peerConnection.getTransceivers().forEach((transceiver) => {
```
Comment (jparez) [inline src/plugins/MediaManager.js:196]: don't know if you can get multiple times the same sender but if not you can use find / findIndex, it's less clearer but the loop stop after getting the first row which match

const arrTransceivers =  this.instance.peerConnection.getTransceivers()
const index = arrTransceivers .findIndex(transceiver => transceiver.sender === sender)
if (index > -1)  arrTransceivers[index].direction = direction

File: src/plugins/MediaManager.js [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+            this.startAudioStreaming();
+        } else {
+            this.stopAudioStreaming();
+        }
+    }
+
+    /**
+     * Initialize and start client webcam video stream.
+     */
+    startVideoStreaming() {
```
Comment (jparez) [inline src/plugins/MediaManager.js:90]: then method is oldschool and can become unreadable as it grow. startVideoStreaming won't return a promise, so if you want to wait for startVideoStreaming in the future you can't without modify startVideoStreaming

i prefere this way

```
async startVideoStreaming() {
     if (!navigator.mediaDevices) {
         return;
     }
     try{
        const mediaStream = navigator.mediaDevices.getUserMedia({
            audio: this.videoWithMicrophone,
            video: {
                width: this.videoWidth,
                height: this.videoHeight,
            },
        })
        log.debug('Client video stream ready');
        this.videoStreaming = true;
        this.localVideoStream = mediaStream;
        this.addVideoStream(mediaStream);
      } catch((err)=> {
        this.onVideoStreamError(err);
      })           
}
```

File: src/plugins/Camera.js [hunk 4]
```diff
@@ -60,73 +58,9 @@ module.exports = class Camera extends OverlayPlugin {
...
-    /**
-     * Toggle local video forward forwarding.
-     * Redirect the client webcam video stream to the instance.
-     */
-    toggleStreaming() {
-        if (!this.streaming) {
-            this.startStreaming();
-        } else {
-            this.stopStreaming();
+            this.toolbarBtn.onclick = () => {
```
Comment (jparez) [inline src/plugins/Camera.js:61]: maybe i wrong but i think 

`this.toolbarBtn.onclick = this.instance.mediaManager.toggleVideoStreaming`

Could get the job. Here we bind an arrow fn which call a fn.

File: src/DeviceRenderer.js [hunk 5]
```diff
@@ -298,8 +299,8 @@ module.exports = class DeviceRenderer {
     disconnect() {
         this.initialized = false;
 
-        if (typeof this.camera !== 'undefined' && typeof this.camera.getClientVideoStream !== 'undefined') {
-            this.removeLocalStream(this.camera.getClientVideoStream());
+        if (typeof this.mediaManager !== 'undefined') {
```
Comment (jparez) [inline src/DeviceRenderer.js:302]: `typeof this.mediaManager !== 'undefined' && this.mediaManager.disconnect();`

Is more sexy, isn't ?

Comment (pgivel) [reply src/DeviceRenderer.js:302]: more sexy but less readable so I'm not sure if it's better in this case :/

File: src/DeviceRenderer.js [hunk 6]
```diff
@@ -40,7 +41,6 @@ module.exports = class DeviceRenderer {
 
         // Websocket
         this.webRTCWebsocket = null;
-        this.webRTCWebsocketName = 'gm-webrtc';
```
Comment (JulienBolard) [inline src/DeviceRenderer.js:43]: It could be preferable to defined it but set it  to null

File: src/plugins/MediaManager.js [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            return false;
+        }
+
+        navigator.permissions.query({name: 'microphone'})
+            .then((permissionObj) => {
+                log.debug(`microphone ${permissionObj.state}`);
+                permissionObj.addEventListener('change', this.onMicrophonePermissionChange.bind(this));
+            })
+            .catch((error) => {
+                log.debug('Can\'t get microphone permission object', error);
```
Comment (JulienBolard) [inline src/plugins/MediaManager.js:48]: this should `return false;`

File: src/plugins/MediaManager.js [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+            });
+
+        navigator.permissions.query({name: 'camera'})
+            .then((permissionObj) => {
+                log.debug(`camera ${permissionObj.state}`);
+                permissionObj.addEventListener('change', this.onCameraPermissionChange.bind(this));
+            })
+            .catch((error) => {
+                log.debug('Can\'t get camera permission object', error);
```
Comment (JulienBolard) [inline src/plugins/MediaManager.js:57]: same here: `return false;`

File: src/plugins/MediaManager.js (hunk 3 above)
Comment (JulienBolard) [reply src/plugins/MediaManager.js:90]: It is really interesting to read that .then() is old school and you propose a more conventional way of writing code. I really don't understand JS dev :)

File: src/plugins/MediaManager.js [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+            });
+    }
+
+    /**
+     * Client video stream error handler.
+     *
+     * @param {Error} error Camera stream error.
+     */
+    onVideoStreamError(error) {
+        log.warn('Can\'t start client video stream', error);
```
Comment (JulienBolard) [inline src/plugins/MediaManager.js:142]: log.error instead ?

File: src/plugins/MediaManager.js [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+    }
+
+    /**
+     * Client audio stream error handler.
+     *
+     * @param {Error} error Microphone stream error.
+     */
+    onAudioStreamError(error) {
+        log.warn('Can\'t start client audio stream', error);
```
Comment (JulienBolard) [inline src/plugins/MediaManager.js:151]: log.error instead ?

File: src/plugins/MediaManager.js (hunk 2 above)
Comment (JulienBolard) [reply src/plugins/MediaManager.js:196]: This needs to be verified as I am not sure if there was multiple instance of the same sender when I did this dev

File: src/plugins/MediaManager.js [hunk 11]
```diff
@@ -0,0 +1,341 @@
...
+            });
+            if (this.microphoneSender) {
+                this.instance.peerConnection.removeTrack(this.microphoneSender);
+            }
+            this.instance.renegotiateWebRTCConnection();
+        }
+    }
+
+    /**
+     * Remove a local video stream and stop sending it through SDP renegotiation.
```
Comment (JulienBolard) [inline src/plugins/MediaManager.js:304]: It also remove the audio track not just the video. Not sure if it is really intended now we split audio and video

**Review by JulienBolard — CHANGES_REQUESTED** [review]

File: src/plugins/MediaManager.js (hunk 3 above)
Comment (pgivel) [reply src/plugins/MediaManager.js:90]: I was also surprised at this but indeed, the browser support versions seems to indicate that async/await + try/catch is much newer than promises

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#browser_compatibility
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise#browser_compatibility

File: src/plugins/MediaManager.js (hunk 11 above)
Comment (pgivel) [reply src/plugins/MediaManager.js:304]: I should add comments to further explain that.

File: tests/unit/camera.test.js [hunk 12]
```diff
@@ -16,12 +17,13 @@ describe('Camera Plugin', () => {
             navbar: true,
             power: true
         });
-        camera = new Camera(instance, false);
+        mediaManager = new MediaManager(instance);
     });
 
     describe('api', () => {
         test('exposes a high level constructor', () => {
             expect(typeof Camera).toBe('function');
```
[the remark this answers is not in the document]
Comment (pgivel) [reply tests/unit/camera.test.js:25]: I'm not sure if this unit test is supposed to test the camera plugin (the ui mostly) or the mediamanager plugin (the functionality). I should probably split it in two tests.

File: src/plugins/MediaManager.js (hunk 2 above)
Comment (pgivel) [reply src/plugins/MediaManager.js:196]: I admit I'm not sure what's the point of setting the direction in the first place. I guess if the transceiver is in the `receive` direction we can't send data through the stream ? But how/why would it be in the `receive` direction?

Comment (pgivel) [reply src/plugins/MediaManager.js:196]: (since I don't understand the aim here, I'm not sure how I can verify that @jparez's suggestion doesn't break stuff)

File: src/DeviceRenderer.js (hunk 6 above)
Comment (pgivel) [reply src/DeviceRenderer.js:43]: You're right, it'll be more legible. But first I'll need to test if setting up the webrtc socket with a `null` protocol is the same as setting it up with an `undefined` one.

Comment (pgivel) [reply src/DeviceRenderer.js:43]: @JulienBolard setting the protocol to `null` doesn't work, the socket is initialized with a `null` protocol instead of without a specific one, and so the connexion isn't established.

File: src/plugins/MediaManager.js (hunk 2 above)
Comment (pgivel) [reply src/plugins/MediaManager.js:198]: closing this suggestion since the existing code works already and I don't know how to check that I'm not breaking stuff by changing it

File: src/plugins/MediaManager.js [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+     *  This flag is used when we don't want to split audio&video and instead want the audio bundled with the video in the same stream.
+     *  Defaults to the microphone flag in the instance options
+     * @param {number} videoWidth Maximum video width. Defaults to 1280
+     * @param {number} videoHeight Maximum video height. Defaults to 720
+     */
+    constructor(instance, videoWithMicrophone = instance.options.microphone, videoWidth = 1280, videoHeight = 720) {
+        this.instance = instance;
+        this.instance.mediaManager = this;
+
+        if (!navigator.mediaDevices || !this.setupPermissions()) {
```
Comment (Previsou) [inline src/plugins/MediaManager.js:20]: no need to await `this.setupPermissions()` ?

File: src/plugins/MediaManager.js [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+            const mediaStream = await navigator.mediaDevices.getUserMedia({
+                audio: this.videoWithMicrophone,
+                video: {
+                    width: this.videoWidth,
+                    height: this.videoHeight,
+                },
+            });
+            log.debug('Client video stream ready');
+            this.videoStreaming = true;
+            this.localVideoStream = mediaStream;
```
Comment (Previsou) [inline src/plugins/MediaManager.js:109]: Should we set audio related fields as well if `this.videoWithMicrophone = true` ?  Or adapt toggleAudioStreaming to protected against starting an audio stream while mic is already captured?

**Review by Previsou — APPROVED** [review]
New to the codebase, so more a cursory review than anything else :-)

File: src/plugins/MediaManager.js (hunk 13 above)
Comment (jparez) [reply src/plugins/MediaManager.js:20]: Yes indeed, but correct me if i'm wrong, constructors cannot be async. 
We can use this.setupPermissions.then(()=>{}) but instance props (this.instance.mediaEventsEnabled, ...) may be initialized after they are used

I guess even if setupPermissions return false, since it's return a promise so !this.setupPermissions will be always true, the test pass.

missed this one :$

**Review by jparez — APPROVED** [review]

File: src/plugins/MediaManager.js (hunk 13 above)
Comment (Previsou) [reply src/plugins/MediaManager.js:20]: > Yes indeed, but correct me if i'm wrong, constructors cannot be async.
I am far from fluent in javascript, so I really don't know. Kinda makes sense though to avoid a promise on a futur object.

File: src/plugins/MediaManager.js (hunk 14 above)
Comment (pgivel) [reply src/plugins/MediaManager.js:110]: hm is it really an issue if the user sends a video&audio stream + an audio stream ?

File: src/plugins/MediaManager.js (hunk 13 above)
Comment (pgivel) [reply src/plugins/MediaManager.js]: Is it better now?

File: README.md [hunk 15]
```diff
@@ -276,7 +276,8 @@ Enables or disables the fullscreen widget. This widget can be used to make the r
 - **Default:** `true`
 - **Compatibility:** `PaaS`, `SaaS`
 - **Details:**
-Enables or disables the camera widget. This widget can be used to forward local webcam to the Android virtual device.
+Enables or disables the camera widget. This widget can be used to forward local webcam video to the Android virtual device.
+By default, if the `microphone` property is also true, then the webcam audio will be used as well.
```
[the remark this answers is not in the document]
Comment (pgivel) [reply README.md:280]: indeed, I'll update that

File: src/plugins/MediaManager.js [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+        this.videoWithMicrophone = videoWithMicrophone;
+    }
+
+    /**
+     * Tries to get the current microphone and camera permission status. If it can, binds listeners to their change
+     * @returns {boolean} True if it could find the permissions object
+     */
+    async setupPermissions() {
+        if (!navigator.permissions) {
+            return false;
```
Comment (Previsou) [inline src/plugins/MediaManager.js:42]: Shouldn't you set `this.instance.mediaEventsEnabled = false` here? In fact, this field seems totally unused

Comment (pgivel) [reply src/plugins/MediaManager.js:42]: `mediaEventsEnabled` is to be used by other classes above this manager (for example for disabling the UI), so that's why it is unused in this class.

It is set to `true` on line 26, as long as we're able to use the `mediaDevices` API. I don't want to set it to false if we can't use the `permissions` API because the feature still works, it simply won't be able to react dynamically to permission changes.

Comment (Previsou) [reply src/plugins/MediaManager.js:42]: Thanks for the explanation

**Review by Previsou — APPROVED** [review]

**Review by JulienBolard — APPROVED** [review]
